x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/RateLimiter.java

```diff
@@ -0,0 +1,144 @@
+    }
+
+    // default for testing
+    RateLimiter(double accumulatedTokensLimit, double tokensPerTimeUnit, TimeUnit unit, Sleeper sleeper, Clock clock) {
+        this.sleeper = Objects.requireNonNull(sleeper);
+        this.clock = Objects.requireNonNull(clock);
+        nextTokenAvailability = Instant.MIN;
+        setRate(accumulatedTokensLimit, tokensPerTimeUnit, unit);
+    }
+
+    public final synchronized void setRate(double newAccumulatedTokensLimit, double newTokensPerTimeUnit, TimeUnit newUnit) {
+        Objects.requireNonNull(newUnit);
+
+        if (newAccumulatedTokensLimit < 0) {
+            throw new IllegalArgumentException("Accumulated tokens limit must be greater than or equal to 0");
+        }
+
+        if (newAccumulatedTokensLimit == Double.POSITIVE_INFINITY) {
+            throw new IllegalArgumentException(
+                Strings.format("Accumulated tokens limit must be less than or equal to %s", Double.MAX_VALUE)
+            );
+        }
+
+        if (newTokensPerTimeUnit <= 0) {
+            throw new IllegalArgumentException("Tokens per time unit must be greater than 0");
+        }
+
+        if (newTokensPerTimeUnit == Double.POSITIVE_INFINITY) {
+            throw new IllegalArgumentException(Strings.format("Tokens per time unit must be less than or equal to %s", Double.MAX_VALUE));
+        }
+
+        accumulatedTokens = Math.min(accumulatedTokens, newAccumulatedTokensLimit);
+
+        accumulatedTokensLimit = newAccumulatedTokensLimit;
+
+        var unitsInNanos = newUnit.toNanos(1);
+        tokensPerNanos = newTokensPerTimeUnit / unitsInNanos;
+        assert tokensPerNanos != Double.POSITIVE_INFINITY : "Tokens per nanosecond should not be infinity";
+
+        accumulateTokens();
+    }
+
+    /**
+     * Causes the thread to wait until the tokens are available
+     * @param tokens the number of items of work that should be throttled, typically you'd pass a value of 1 here
+     * @throws InterruptedException
+     */
+    public void acquire(int tokens) throws InterruptedException {
+        if (tokens <= 0) {
+            throw new IllegalArgumentException("Requested tokens must be positive");
+        }
+
+        double nanosToWait;
+        synchronized (this) {
+            accumulateTokens();
+            var accumulatedTokensToUse = Math.min(tokens, accumulatedTokens);
+            var additionalTokensRequired = tokens - accumulatedTokensToUse;
+            nanosToWait = additionalTokensRequired / tokensPerNanos;
+            accumulatedTokens -= accumulatedTokensToUse;
+            nextTokenAvailability = nextTokenAvailability.plus(Duration.ofNanos((long) nanosToWait));
+        }
+
+        sleeper.sleep((long) nanosToWait);
+    }
+
+    private void accumulateTokens() {
+        var now = Instant.now(clock);
+        if (now.isAfter(nextTokenAvailability)) {
+            var elapsedTimeNanos = nanosBetweenExact(nextTokenAvailability, now);
```

Comment:
This is called in the ctor via `setRate()` at which point `nextTokenAvailability == Instant.MIN`. Because the calculated `elapsedTimeNanos` is high the class will be initialised with `accumulatedTokens ==  accumulatedTokensLimit`. 

That seems reasonable to me, or at least as good as initialising `accumulatedTokens` to 0. Just want to check that is the intention

Reply:
Yeah that was intentionally. My thinking was that the first request can move forward without having to wait for tokens to accumulate if the limit was set to a positive number. If we always want it to start as 0 that's fine with me too though.